Implement statistic pad mode in jax.numpy.pad

minoring opened the pull request:
Implemented maximum, mean, median, and minimum padding modes.

Related #5010

minoring committed:
Implement statistic pad mode in jax.numpy.pad

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -2342,6 +2342,23 @@ def _pad_edge(array, pad_width):
+
+    padded_values = stat_func(array, axis=i, keepdims=True)
+    if np.issubdtype(array.dtype, np.integer):
+      padded_values = jax.numpy.round(padded_values)
```
`round` is defined in this file, so you can drop `jax.numpy` here.

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -2350,6 +2367,9 @@ def _pad(array, pad_width, mode, constant_values):
     return array
 
+  stat_funcs = {"maximum": jax.numpy.amax, "minimum": jax.numpy.amin,
+                "mean": jax.numpy.mean, "median": jax.numpy.median}
```
Same with these: all these aggregations are defined directly in this file so you can drop `jax.numpy`.

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -2342,6 +2342,23 @@ def _pad_edge(array, pad_width):
   return array
 
 
+def _pad_stats(array, pad_width, stat_func):
```
Have you thought about handling the optional `stat_length` argument? If you'd prefer, feel free to leave it as a TODO.

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -1306,6 +1306,10 @@ def testOperatorRound(self):
+          ('maximum', nonempty_shapes),
+          ('minimum', nonempty_shapes),
+          ('mean', nonempty_shapes),
+          ('median', nonempty_shapes),
```
I'd lean toward splitting these into separate tests, because `constant_values` has no effect on these modes, and so this parametrization is generating ~10 identical tests for each. This would also allow separate parametrization over the `stat_length` argument, if/when it is implemented.

jakevdp reviewed (changes requested):
Looks great! Thanks for tackling this. A few comments below.

minoring commented:
@jakevdp Thank you for your kind review.
I updated PR. Please take a look :)

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -2342,35 +2342,82 @@ def _pad_edge(array, pad_width):
+      if length_before > array_length:
+        length_before = array_length
+      if length_after > array_length:
+        length_after = array_length
```
nit: maybe use `length_before = mlin(length_before, array_length)`

minoring replied:
I tried `length_before = min(length_before, array_length)` and got an error.
I suspect `jax.numpy.min` shadows python function `min`

jakevdp replied:
Sorry, yes you need to use `_min()` here to get the builtin `min()`.

minoring replied:
Oh I see! updated ;)

jakevdp commented on jax/_src/numpy/lax_numpy.py:
```diff
@@ -2342,35 +2342,82 @@ def _pad_edge(array, pad_width):
+
+      slice_before = lax.slice_in_dim(array, 0, length_before, axis=i)
+      slice_after = lax.slice_in_dim(array, array_length - length_after,
+                                     array_length, axis=i)
```
More concise, I think this works: `slice_after = lax.slice_in_dim(array, -length_after, None, axis=i)`

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -1351,6 +1351,54 @@ def testPad(self, shape, dtype, mode, pad_width, constant_values, rng_factory):
+          ('mean', nonempty_shapes),
+          ('median', nonempty_shapes),
+      ]
+      for shape, dtype in _shape_and_dtypes(shapes, all_dtypes)
```
since all shapes here are the same, you could simplify this and pass `nonempty_shapes` directly here.

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -1351,6 +1351,54 @@ def testPad(self, shape, dtype, mode, pad_width, constant_values, rng_factory):
+          # pad
+          0, 1,
+      ]
+      for stat_length in [
```
We should probably test `None` as well.

jakevdp commented on tests/lax_numpy_test.py:
```diff
@@ -1351,6 +1351,54 @@ def testPad(self, shape, dtype, mode, pad_width, constant_values, rng_factory):
+      ]
+      if pad_width != ()))
+  def testPadStatValues(self, shape, dtype, mode, pad_width, stat_length, rng_factory):
+    rng = rng_factory(self.rng())
```
Minor: recently, we've started using `jtu.rand_default()` directly to avoid having to add `rng_factory` to all the parametrization boilerplate.

jakevdp reviewed (changes requested):
Great progress! Nice work on the updated test - it has great coverage of all the various valid inputs.

A few mostly minor comments to address, and then we can get this merged!

minoring commented:
Nice tips. Thank you!
I made some improvement. PTAL :)

jakevdp reviewed (approved):
Looks great, thanks! I'm going to run this through a fuller set of tests (including GPU & TPU) and see if we can get this merged.

jakevdp commented:
OK, so there are a number of issues that came up in more comprehensive testing:

- passing `stat_length=()` for mode in (median, min, max) results in `ValueError: zero-size array to reduction operation minimum which has no identity`. It might be that an error is expected in this case: I'm not sure. But we need to shore-up the tests for this either way.
- with `dtype=bool`, `stat_length=2` and `mode='mean'`, there are differences on TPU (possibly because of differences in how `0.5` is converted to boolean, but I'm not certain). I think it would be fine to leave boolean dtypes out of these tests... computing means of booleans is not particularly well defined anyway.

Would you mind taking a look at those? Thanks!

minoring commented:
- Removed `stat_length=()` from tests since numpy gave me the same error message.
- Removed calculating mean of booleans. I considered leaving whole boolean type out of tests, but computing others might be OK since boolean can act like integer. https://docs.python.org/3/library/stdtypes.html#boolean-values

jakevdp commented:
OK, I'll check the tests again.

Last thing: can you please squash all the commits to a single commit and force-push to this branch? If that's not something you're comfortable doing, I can walk you through it or take care of it for you. Thanks!

minoring commented:
Squashed it into single commit. 👍

jakevdp commented:
Thanks! Tests look good – I'll work on getting this merged!

jakevdp commented:
We have some internal test failures that I believe would be remedied by syncing to master and force-pushing to this branch. Are you comfortable doing that?

minoring commented:
Syncing to master worked! Thanks.
